torchrl/envs/transforms/transforms.py

```diff
@@ -6796,3 +6795,211 @@ def _reset(
+            tensor dictionary. Defaults to "reward".
+        reward_function (Optional[callable]): The reward function to use for calculating
+            the rewards of augmented samples. Defaults to None, in which case the
+            `distance_reward_function` is used.
+
+    Attributes:
+        ENV_ERR (str): The error message to raise when the transform is applied to
+            the collector or the environment.
+
+    """
+
+    ENV_ERR = (
+        "The Reward2GoTransform is only an inverse transform and can "
+        "only be applied to the replay buffer and not to the collector or the environment."
+    )
+
+    def __init__(
+        self,
+        samples: Optional[Union[int, torch.Tensor]] = 4,
+        generation_type: str = "future",
+        achieved_goal_key: Optional[NestedKey] = "achieved_goal",
+        desired_goal_key: Optional[NestedKey] = "desired_goal",
+        reward_key: Optional[NestedKey] = "reward",
+        reward_function: Optional[callable] = None,
+    ):
+        super().__init__(
+            in_keys=None,
+            in_keys_inv=None,
+            out_keys_inv=None,
+        )
+        self.achieved_goal_key = achieved_goal_key
+        self.desired_goal_key = desired_goal_key
+        self.reward_key = reward_key
+        self.generation_type = generation_type
+
+        if reward_function is None:
+            self.reward_function = distance_reward_function
+        else:
+            self.reward_function = reward_function
+
+        if not isinstance(samples, torch.Tensor):
+            samples = torch.tensor(samples)
+
+        self.register_buffer("samples", samples)
+
+    def _inv_call(self, tensordict: TensorDictBase) -> TensorDictBase:
+        augmentation_td = self.her_augmentation(tensordict)
+        return torch.cat([tensordict, augmentation_td], dim=0)
+
+    def _inv_apply_transform(self, tensordict: TensorDictBase) -> torch.Tensor:
+        return self.her_augmentation(tensordict)
+
+    def forward(self, tensordict: TensorDictBase) -> TensorDictBase:
+        return tensordict
+
+    def _call(self, tensordict: TensorDictBase) -> TensorDictBase:
+        raise ValueError(self.ENV_ERR)
+
+    def her_augmentation(self, sampled_td: TensorDictBase):
+        if len(sampled_td.shape) == 1:
+            sampled_td = sampled_td.unsqueeze(0)
+        b, t = sampled_td.shape
+        trajectories = _get_num_per_traj(sampled_td.get("terminated"))
+        splitted_td = _split_and_pad_sequence(sampled_td, trajectories)
+        splitted_achieved_goals = splitted_td.get(self.achieved_goal_key)
+
+        # get indices for each trajectory
+        idxs = self.generate_sample_idxs(trajectories)
+
+        # create new goals based idxs
+        new_goals = []
+        for i, ids in enumerate(idxs):
+            new_goals.append(splitted_achieved_goals[i][ids])
+
+        # calculate rewards given new desired goals and old achieved goals
+        vmap_rewards = torch.vmap(distance_reward_function)
```

Comment:
I think you wanna call ```self.reward_function``` instead of ```distance_reward_function```. Also maybe the ```reward_function``` should be a TensorDictModule such it can be more easily customized for a given environment. There is a ```torchrl.modules.VmapModule``` for wrapping TensorDictModules with with vmap.